File: backend/app/schemas/workspace.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class WorkspaceSessionResponse(BaseModel):
    workspace_id: str
    expires_at: datetime
    hosted_mode: bool = True
    csrf_required: bool = True
    llm_optional: bool = True
    session_token: str | None = None
    csrf_token: str | None = None


class HostedUserProfile(BaseModel):
    id: str
    email: str
    full_name: str
    avatar_url: str = ""
    auth_provider: str = "supabase"
    email_verified: bool = False


class HostedWorkspaceSummary(BaseModel):
    id: str
    name: str
    slug: str
    role: str = "owner"
    plan: str = "free"
    subscription_status: str = "inactive"


class HostedFeatureFlags(BaseModel):
    platform_managed_ai: bool = True
    runtime_llm_configurable: bool = False
    billing_enabled: bool = False


class HostedBootstrapResponse(BaseModel):
    hosted_mode: bool = True
    auth_required: bool = True
    csrf_required: bool = False
    llm_optional: bool = True
    user: HostedUserProfile
    workspace: HostedWorkspaceSummary
    features: HostedFeatureFlags = Field(default_factory=HostedFeatureFlags)


class PlaceSelection(BaseModel):
    label: str
    kind: Literal["city", "region", "country", "manual"] = "manual"
    match_scope: Literal["city", "metro", "region", "country"] = "city"
    city: str = ""
    region: str = ""
    country: str = ""
    country_code: str = ""
    lat: float | None = None
    lon: float | None = None
    provider: str = "manual"
    provider_id: str = ""


class CompensationPreference(BaseModel):
    currency: str = "USD"
    pay_period: Literal["hourly", "monthly", "annual"] = "annual"
    current_comp: float | None = None
    min_base: float | None = None
    target_total_comp: float | None = None
    min_acceptable_tc: float | None = None
    include_equity: bool = True


class WorkspacePreferences(BaseModel):
    roles: list[str] = Field(default_factory=list)
    keywords: list[str] = Field(default_factory=list)
    companies: list[str] = Field(default_factory=list)
    preferred_places: list[PlaceSelection] = Field(default_factory=list)
    workplace_preference: Literal["remote_friendly", "remote_only", "location_only"] = "remote_friendly"
    max_days_old: int = 14
    include_linkedin_jobs: bool = False
    current_title: str = ""
    current_level: str = "mid"
    compensation: CompensationPreference = Field(default_factory=CompensationPreference)
    exclude_staffing_agencies: bool = True


class WorkspaceResumeStatus(BaseModel):
    exists: bool = False
    filename: str = ""
    file_size: int = 0
    parse_status: Literal["missing", "parsed", "warning", "error"] = "missing"
    parse_warning: str = ""


class OnboardingState(BaseModel):
    workspace_id: str
    has_started_search: bool = False
    needs_resume: bool = True
    needs_preferences: bool = True
    ready_to_search: bool = False
    resume_warning: str = ""
    llm_optional: bool = True
    llm_available: bool = False
    resume: WorkspaceResumeStatus = Field(default_factory=WorkspaceResumeStatus)
    preferences: WorkspacePreferences = Field(default_factory=WorkspacePreferences)


class WorkspaceResumeUploadResponse(BaseModel):
    message: str = "Resume uploaded successfully"
    resume: WorkspaceResumeStatus
    analysis: dict | None = None


class LocationSuggestion(BaseModel):
    label: str
    kind: Literal["city", "region", "country", "manual"] = "manual"
    match_scope: Literal["city", "metro", "region", "country"] = "city"
    subtitle: str = ""
    city: str = ""
    region: str = ""
    country: str = ""
    country_code: str = ""
    lat: float | None = None
    lon: float | None = None
    provider: str = "manual"
    provider_id: str = ""


class SearchSnapshot(BaseModel):
    roles: list[str] = Field(default_factory=list)
    keywords: list[str] = Field(default_factory=list)
    companies: list[str] = Field(default_factory=list)
    preferred_places: list[PlaceSelection] = Field(default_factory=list)
    workplace_preference: Literal["remote_friendly", "remote_only", "location_only"] = "remote_friendly"
    max_days_old: int = 14
    include_linkedin_jobs: bool = False
    current_title: str = ""
    current_level: str = ""
    compensation: CompensationPreference = Field(default_factory=CompensationPreference)
    exclude_staffing_agencies: bool = True


class GeneratedProfileResponse(BaseModel):
    """LLM-tailored search profile generated from a workspace's resume.

    Wire-compat with the GeneratedProfile pydantic model in
    src/job_finder/models/schemas.py — same field names so the frontend
    can parse one shape regardless of which layer produced it.

    Returned by POST /onboarding/generate-profile. The frontend uses this
    to populate the ReadyToLaunchHero with an AI-tailored alternative to
    the hardcoded archetype templates, so users in niches we never
    modeled (climate tech, vet med, MTS, web3, etc.) get a custom
    profile from their resume instead of being forced into a preset.
    """

    detected_archetype: str
    confidence: float
    reasoning: str
    closest_template: str | None = None
    career_target: str
    seniority_signal: str
    scoring: dict[str, float]
    keywords: dict[str, list[str]]
    target_roles: list[str] = Field(default_factory=list)
    compensation: dict
    enabled_scrapers: list[str] = Field(default_factory=list)
    recommended_external_boards: list[str] = Field(default_factory=list)
    primary_strengths: list[str] = Field(default_factory=list)
    development_areas: list[str] = Field(default_factory=list)
    # Set by the API layer when the result was served from cache.
    cached: bool = False


class WorkspaceSearchRunResponse(BaseModel):
    run_id: str
    status: Literal["pending", "running", "completed", "failed"]
    started_at: datetime | None = None
    completed_at: datetime | None = None

File: frontend/src/types/workspace.ts
export interface WorkspaceSession {
  workspace_id: string;
  expires_at: string;
  hosted_mode: boolean;
  csrf_required: boolean;
  llm_optional: boolean;
  session_token?: string | null;
  csrf_token?: string | null;
}

export interface HostedUserProfile {
  id: string;
  email: string;
  full_name: string;
  avatar_url: string;
  auth_provider: string;
  email_verified: boolean;
}

export interface HostedWorkspaceSummary {
  id: string;
  name: string;
  slug: string;
  role: string;
  plan: string;
  subscription_status: string;
}

export interface HostedFeatureFlags {
  platform_managed_ai: boolean;
  runtime_llm_configurable: boolean;
  billing_enabled: boolean;
}

export interface DevHostedPersona {
  id: string;
  email: string;
  full_name: string;
  avatar_url: string;
  headline: string;
  background: string;
  job_search_focus: string;
  current_title: string;
  current_level: string;
  target_roles: string[];
  keywords: string[];
  preferred_places: string[];
  workplace_preference: 'remote_friendly' | 'remote_only' | 'location_only';
  resume_filename: string;
}

export interface DevHostedLoginResponse {
  access_token: string;
  token_type: 'bearer';
  expires_at: string;
  persona: DevHostedPersona;
}

export interface DevHostedUser {
  id: string;
  email: string;
  full_name: string;
  avatar_url: string;
  auth_provider: string;
  seeded: boolean;
}

export interface DevHostedRegisterResponse {
  access_token: string;
  token_type: 'bearer';
  expires_at: string;
  user: DevHostedUser;
}

export interface HostedBootstrap {
  hosted_mode: true;
  auth_required: true;
  csrf_required: false;
  llm_optional: boolean;
  user: HostedUserProfile;
  workspace: HostedWorkspaceSummary;
  features: HostedFeatureFlags;
}

export interface PlaceSelection {
  label: string;
  kind: 'city' | 'region' | 'country' | 'manual';
  match_scope: 'city' | 'metro' | 'region' | 'country';
  city: string;
  region: string;
  country: string;
  country_code: string;
  lat: number | null;
  lon: number | null;
  provider: string;
  provider_id: string;
}

export interface CompensationPreference {
  currency: string;
  pay_period: 'hourly' | 'monthly' | 'annual';
  current_comp: number | null;
  min_base: number | null;
  target_total_comp: number | null;
  min_acceptable_tc: number | null;
  include_equity: boolean;
}

export interface WorkspacePreferences {
  roles: string[];
  keywords: string[];
  companies: string[];
  preferred_places: PlaceSelection[];
  workplace_preference: 'remote_friendly' | 'remote_only' | 'location_only';
  max_days_old: number;
  include_linkedin_jobs: boolean;
  current_title: string;
  current_level: string;
  compensation: CompensationPreference;
  exclude_staffing_agencies: boolean;
}

export interface WorkspaceResumeStatus {
  exists: boolean;
  filename: string;
  file_size: number;
  parse_status: 'missing' | 'parsed' | 'warning' | 'error';
  parse_warning: string;
}

export interface OnboardingState {
  workspace_id: string;
  has_started_search: boolean;
  needs_resume: boolean;
  needs_preferences: boolean;
  ready_to_search: boolean;
  resume_warning: string;
  llm_optional: boolean;
  llm_available: boolean;
  resume: WorkspaceResumeStatus;
  preferences: WorkspacePreferences;
}

export interface WorkspaceResumeUploadResponse {
  message: string;
  resume: WorkspaceResumeStatus;
  analysis: Record<string, unknown> | null;
}

export interface WorkspaceSearchRunResponse {
  run_id: string;
  status: 'pending' | 'running' | 'completed' | 'failed';
  started_at: string | null;
  completed_at: string | null;
}

/**
 * LLM-tailored search profile generated from the workspace's resume.
 *
 * The whole point of this object is to make Launchboard work for *any*
 * career, not just the seven hardcoded archetype templates. A user with
 * an unusual or multi-domain background gets a profile generated
 * specifically for them — covering niches we never modeled (climate
 * tech, vet med, MTS at frontier labs, biotech regulatory, etc.).
 *
 * Wire-compat with the backend GeneratedProfileResponse and the
 * pipeline-layer GeneratedProfile pydantic model — same field names
 * top to bottom.
 */
export interface GeneratedProfileScoring {
  technical_skills: number;
  leadership_signal: number;
  career_progression: number;
  platform_building: number;
  comp_potential: number;
  company_trajectory: number;
  culture_fit: number;
}

export interface GeneratedProfileKeywords {
  technical: string[];
  leadership: string[];
  signal_terms: string[];
}

export interface GeneratedProfileCompensation {
  currency: string;
  pay_period: string;
  min_base: number;
  target_total_comp: number;
  include_equity: boolean;
}

export interface GeneratedProfile {
  detected_archetype: string;
  confidence: number;
  reasoning: string;
  closest_template: string | null;
  career_target: string;
  seniority_signal: string;
  scoring: GeneratedProfileScoring;
  keywords: GeneratedProfileKeywords;
  target_roles: string[];
  compensation: GeneratedProfileCompensation;
  enabled_scrapers: string[];
  recommended_external_boards: string[];
  primary_strengths: string[];
  development_areas: string[];
  cached: boolean;
}

export interface LocationSuggestion {
  label: string;
  kind: 'city' | 'region' | 'country' | 'manual';
  match_scope: 'city' | 'metro' | 'region' | 'country';
  subtitle: string;
  city: string;
  region: string;
  country: string;
  country_code: string;
  lat: number | null;
  lon: number | null;
  provider: string;
  provider_id: string;
}
